src/main/java/net/dv8tion/jda/core/requests/WebSocketClient.java

```diff
@@ -283,136 +281,7 @@ protected boolean send(String message, boolean skipQueue)
-                            audioRequest.setNextAttemptEpoch(System.currentTimeMillis() + 2000);
-                            //If we are already in the correct state according to voice state
-                            // we will not receive a VOICE_STATE_UPDATE that would remove it
-                            // thus we update it here
-                            final GuildVoiceState voiceState = guild.getSelfMember().getVoiceState();
-                            updateAudioConnection0(guild.getIdLong(), voiceState.getChannel());
-                        }
-                        maybeUnlock();
-                        attemptedToSend = true;
-                    }
-                    else
-                    {
-                        String message = ratelimitQueue.peek();
-                        if (message != null)
-                        {
-                            needRatelimit = !send(message, false);
-                            if (!needRatelimit)
-                                ratelimitQueue.remove();
-                            attemptedToSend = true;
-                        }
-                    }
-
-                    if (needRatelimit || !attemptedToSend)
-                        Thread.sleep(1000);
-                }
-                catch (InterruptedException ignored)
-                {
-                    LOG.debug("Main WS send thread interrupted. Most likely JDA is disconnecting the websocket.");
-                    break;
-                }
-                finally
-                {
-                    // on any exception that might cause this lock to not release
-                    maybeUnlock();
-                }
-            }
-        });
-        ratelimitThread.setUncaughtExceptionHandler((thread, throwable) ->
-        {
-            handleCallbackError(socket, throwable);
-            setupSendingThread();
-        });
-        ratelimitThread.setName(api.getIdentifierString() + " MainWS-Sending Thread");
-        ratelimitThread.start();
-    }
-
-    protected JSONObject newVoiceClose(long guildId)
-    {
-        return new JSONObject()
-            .put("op", WebSocketCode.VOICE_STATE)
-            .put("d", new JSONObject()
-                .put("guild_id", Long.toUnsignedString(guildId))
-                .put("channel_id", JSONObject.NULL)
-                .put("self_mute", false)
-                .put("self_deaf", false));
-    }
-
-    protected JSONObject newVoiceOpen(AudioManager manager, long channel, long guild)
-    {
-        return new JSONObject()
-            .put("op", WebSocketCode.VOICE_STATE)
-            .put("d", new JSONObject()
-                .put("guild_id", guild)
-                .put("channel_id", channel)
-                .put("self_mute", manager.isSelfMuted())
-                .put("self_deaf", manager.isSelfDeafened()));
+        ratelimitThread = executor.scheduleAtFixedRate(new WebSocketSendingThread(this), 0, 500, TimeUnit.MILLISECONDS);
```

Comment:
This is not entirely accurate. We try to send as fast as possible, only sleeping when:
1) We havent sent auth info
2) We didn't try to send a packet during the latest iteration

By only running once every 500ms, we would take close to a minute to fully connect to 100 voice channels if we were reconnecting.

Reply:
We could change it to re-schedule itself depending on needed timeout, however we still need some kind of idle time when nothing was sent. How about:

- Nothing happened -> wait 500ms
- Sent message -> wait 10ms
- Rate-limit -> wait 60s

Reply:
Since we plan to share threads its fair to give other sessions a window to use it, even this small